client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java

```diff
@@ -1025,6 +1111,35 @@ protected void removeTenantClient() {
         });
     }
 
+    @Override
+    public Future<AsyncCommandClient> getOrCreateAsyncCommandClient(final String tenantId, final String deviceId) {
```

Comment:
I am still struggling with this one. Should it be allowed to have async and non-async command clients for same tenant and device in parallel? I yes, we need somehow to distinguish them in `activeSenders` map etc. or duplicate the code an put it into separate map. I also don't like the casting. Any ideas?

Reply:
Probably using separate `creationRequests` and `creationLocks` for senders and request response clients could resolve this issue

Reply:
I  do not quite see the problem with creating consumers. As for the senders, we could simply prefix the keys used in the activeSenders and creationLocks with the simple class name of the sender to create, couldn't we?

Reply:
This would work. I just had some concerns because we are loosing type safety in some degree. But if you are fine with it I will provide the updated implementation soon.